amethyst_renderer/src/pipe/pass.rs

```diff
@@ -3,174 +3,60 @@
-    }
-}
+pub(crate) struct Description(Arc<Pass>);
 
-impl Debug for ModelPass {
-    fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
-        fmt.debug_tuple("ModelPass")
-            .field(&"[closure]")
-            .field(&self.1)
-            .finish()
+impl Description {
+    pub fn new<P: Pass + 'static>(pass: P) -> Self {
+        Description(Arc::new(pass))
     }
-}
 
-impl LightPass {
-    /// Applies the rendering pass using the given `Encoder`, `Target`, `Scene` and `Light`.
-    pub fn apply(&self, enc: &mut Encoder, out: &Target, scene: &Scene, light: &Light) {
-        (self.0)(enc, out, &self.1, scene, light)
+    pub fn compile(self, fac: &mut Factory, out: &Target) -> Result<CompiledPass> {
+        let eb = NewEffect::new(fac, out);
+        let effect = self.0.compile(eb)?;
+        Ok(CompiledPass {
+            effect,
+            inner: self.0,
+        })
     }
 }
 
-impl Debug for LightPass {
+impl Debug for Description {
     fn fmt(&self, fmt: &mut Formatter) -> FmtResult {
-        fmt.debug_tuple("LightPass")
-            .field(&"[closure]")
-            .field(&self.1)
+        fmt.debug_tuple("PassDesc")
+            .field(&"[impl]")
             .finish()
     }
 }
 
 #[derive(Clone)]
-pub enum PassBuilder<'a> {
-    Basic(BasicFn),
-    Simple(SimpleFn, EffectBuilder<'a>),
-    Model(ModelFn, EffectBuilder<'a>),
-    Light(LightFn, EffectBuilder<'a>),
+pub struct CompiledPass {
+    effect: Effect,
+    inner: Arc<Pass>,
 }
 
-impl<'a> PassBuilder<'a> {
-    pub fn basic<F>(func: F) -> Self
-        where F: Fn(&mut Encoder, &Target) + Send + Sync + 'static
-    {
-        PassBuilder::Basic(Arc::new(func))
-    }
-
-    pub fn simple<F>(eb: EffectBuilder<'a>, func: F) -> Self
-        where F: Fn(&mut Encoder, &Target, &Effect, &Scene) + Send + Sync + 'static
-    {
-        PassBuilder::Simple(Arc::new(func), eb)
-    }
-
-    pub fn model<F>(eb: EffectBuilder<'a>, func: F) -> Self
-        where F: Fn(&mut Encoder, &Target, &Effect, &Scene, &Model) + Send + Sync + 'static
-    {
-        PassBuilder::Model(Arc::new(func), eb)
-    }
-
-    pub fn light<F>(eb: EffectBuilder<'a>, func: F) -> Self
-        where F: Fn(&mut Encoder, &Target, &Effect, &Scene, &Light) + Send + Sync + 'static
-    {
-        PassBuilder::Light(Arc::new(func), eb)
-    }
-
-    pub(crate) fn finish(self, fac: &mut Factory, t: &Targets, out: &Target) -> Result<Pass> {
-        match self {
-            PassBuilder::Basic(f) => Ok(Pass::Basic(BasicPass(f))),
-            PassBuilder::Simple(f, e) => Ok(Pass::Simple(SimplePass(f, e.finish(fac, out)?))),
-            PassBuilder::Model(f, e) => Ok(Pass::Model(ModelPass(f, e.finish(fac, out)?))),
-            PassBuilder::Light(f, e) => Ok(Pass::Light(LightPass(f, e.finish(fac, out)?))),
-        }
+impl CompiledPass {
+    pub fn apply(&self, enc: &mut Encoder, scene: &Scene, model: &Model) {
+        self.inner.apply(enc, &mut self.effect.clone(), scene, model);
```

Comment:
Definitely noted! I need to find a nice way to avoid this clone and remain thread-safe.